bundles/admin/admin-hierarchical-layerlist/components/Layer.js

```diff
@@ -0,0 +1,246 @@
+        });
+        // setup templates for layer types/require only ones supported
+        _.each(this.supportedTypes, function(type) {
+            if (type.header === false) {
+                return;
+            }
+            var file = 'text!_bundle/templates/layer/' + type.id + 'SettingsTemplateHeader.html';
+            require([file], function(header) {
+                type.headerTemplate = _.template(header);
+            }, function() {
+                me.log.warn('No admin header template for layertype: ' + type.id + " file was: " + file);
+            });
+        });
+        _.each(this.supportedTypes, function(type) {
+            if (type.footer === false) {
+                return;
+            }
+            var file = 'text!_bundle/templates/layer/' + type.id + 'SettingsTemplateFooter.html';
+            require([file], function(footer) {
+                type.footerTemplate = _.template(footer);
+            }, function() {
+                me.log.warn('No admin footer template for layertype: ' + type.id + " file was: " + file);
+            });
+        });
+    },
+
+    /**
+     * Get mapl layer groups for layer group selection
+     * @method  _getMaplayerGroups
+     * @return  {Array}           groups
+     * @private
+     */
+    _getMaplayerGroups: function() {
+        var me = this;
+        var groups = [];
+        me.layerService.getAllLayerGroups().forEach(function(group) {
+            groups.push({
+                id: group.id,
+                cls: 'group',
+                name: Oskari.getLocalized(group.name)
+            });
+
+            // subgroups
+            group.groups.forEach(function(subgroup) {
+                groups.push({
+                    id: subgroup.id,
+                    cls: 'subgroup',
+                    name: Oskari.getLocalized(subgroup.name)
+                });
+
+                // subgroup subgroups
+                subgroup.groups.forEach(function(subgroupsubgroup) {
+                    groups.push({
+                        id: subgroupsubgroup.id,
+                        cls: 'subgroupsubgroup',
+                        name: Oskari.getLocalized(subgroupsubgroup.name)
+                    });
+                });
+            });
+        });
+
+        return groups;
+    },
+
+    /*******************************************************************************************************************************
+    /* PUBLIC METHODS
+    *******************************************************************************************************************************/
+    /**
+     * Get dataproviders
+     * @method getDataproviders
+     * @param  {Boolean}        emptyCache empty cache
+     * @param  {Function}       callback   callback function
+     */
+    getDataproviders: function(emptyCache, callback) {
+        var me = this;
+        if (me.dataProviders === null || emptyCache === true) {
```

Comment:
I would reverse this with:

    if(me.dataProviders !== null && !emptyCache) {
        callback(me.dataProviders);
        return;
    }

This way you can get rid of one indentation level for the rest of the function and the if-statement is easier to follow with just 2 lines of code inside it rather than 20.